utils/publish-snapshot.sh

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env bash
+
+# https://www.gnu.org/software/bash/manual/bash.html#The-Set-Builtin
+set -e # Exit immediately if a command exits with a non-zero status.
+set -o pipefail
+set -u # Treat unset variables as an error when substituting.
+
+# This script uses the `changeset` tool to version each package as a snapshot
+# release and then publishes that release to npm with a tag named after the PR.
+#
+# It is designed to be run in a Github Action, and by default will abort if the
+# `CI`, `GITHUB_EVENT_NAME`, or `GITHUB_REF` environment variables are unset.
+
+# Identifies the path that the script is in (http://stackoverflow.com/a/246128/11807)
+MYPATH=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+
+# ROOT is the root directory of our project.
+ROOT="$MYPATH/.."
+
+pushd "$ROOT"
+
+if [ -z ${CI+CI_UNSET} ]; then
+    echo "CI environment variable is unset. Exiting!"
+    exit 1
+fi
+
+if [ -z ${GITHUB_EVENT_NAME+GITHUB_EVENT_NAME_UNSET} ]; then
+    echo "GITHUB_EVENT_NAME environment variable is unset. Exiting!"
+    exit 1
+fi
+
+if [ -z ${GITHUB_REF+GITHUB_REF_UNSET} ]; then
+    echo "GITHUB_REF environment variable is unset. Exiting!"
+    exit 1
+fi
+
+if [[
+    "$GITHUB_EVENT_NAME" != "workflow_dispatch" \
+    && "$GITHUB_EVENT_NAME" != "pull_request"
+]]; then
+    exit
+fi
+
+# Check if we need to do any work
+temp_file=$(mktemp)
+yarn changeset status --verbose --output "$temp_file"
+jq -e '.releases | length | if . > 0 then . else "Error: No changesets found\\n" | halt_error(1) end' < "$temp_file"
+
+
+echo "Running for $GITHUB_EVENT_NAME @ $GITHUB_REF"
+
+# Example GITHUB_REF
+# refs/pull/:prNumber/merge
+if [[ "$GITHUB_REF" =~ refs/pull/([[:digit:]]+)/merge ]]; then
+    echo "Found PR #${BASH_REMATCH[1]}"
+    PR_NUMBER="PR${BASH_REMATCH[1]}"
+else
+    echo "Pull Request number not found in ref. Exiting!"
+    exit 1
+fi
+
+# publish:ci steps
+node "$ROOT/utils/pre-publish-check-ci.js"
+
+if ! git diff --stat --exit-code HEAD; then
+    echo "Git repo is dirty. Stash/Commit your changes before running."
+    exit 1
+fi
```

Comment:
Are you expecting people to run this script locally as well?  If not, you may want to tweak this error message a bit since people won't be able to stash/commit if the repo is dirty when this runs on CI.